src/librbd/api/Trash.cc

```diff
@@ -295,14 +293,18 @@ int Trash<I>::move(librados::IoCtx &io_ctx, rbd_trash_image_source_t source,
     if (r < 0) {
       return r;
     }
-
-    std::experimental::erase_if(
-      trash_image_specs, [image_name](const auto& pair) {
-        const auto& spec = pair.second;
-        return (spec.source != cls::rbd::TRASH_IMAGE_SOURCE_USER ||
-                spec.state != cls::rbd::TRASH_IMAGE_STATE_MOVING ||
-                spec.name != image_name);
-      });
+    // should use std::erase_if() in C++20
+    for (auto it = trash_image_specs.begin(), last = trash_image_specs.end();
+         it != last;) {
```

Comment:
Why are you capturing `last`?  It is going to crash, as `trash_image_specs.end()` is modified.

Reply:
i don't think it's going to crash. see https://en.cppreference.com/w/cpp/container/map/end . `end()` is but a theoretical element which is located after the last element, which does not point to any element.

Reply:
```c++
// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
// vim: ts=8 sw=2 smarttab

#include <map>
#include <iostream>

using namespace std;
int main()
{
  map<int, int> m = {{1,1}, {2,2}};
  for (auto it = m.begin(), last = m.end();
       it != last;) {
    if (const auto& spec = it->second;
    spec == 1) {
      it = m.erase(it);
    } else {
      ++it;
    }
  }
  for (auto [i, j] : m) {
    cout << i << " " << j << endl;
  }
}
```

i created a mini reproducer. could give it a try.

Reply:
Looks like it indeed doesn't crash for a map, but it would crash if `m` was a vector.  And it certainly stands out.

What is the benefit of capturing `last`?  Do you assume `m.end()` to be expensive?

Reply:
i am trying to replicate the implementation of libc++ and libstdc++, and don't want to take any risk of performance regression caused by my change. they basically share the same implementation. for the fun of reading: https://github.com/llvm-mirror/libcxx/blob/78d6a7767ed57b50122a161b91f59f19c9bd0d19/include/functional#L3083-L3092.

Reply:
It is not an IO path so no worries about performance. The code clarity is only important.

Saying this I am fine with the current version, though personally I would just compare with `end()`, as I used to. This `last` thing just makes me stumble when reading code, trying to figure out if there is no some additional effect.

Reply:
Exactly.   Let's go with a classic `it != m.end()`.

Reply:
@trociny @idryomov  thanks for your insights. i wanted to minimize the impact of this change, but i also want to minimize the frequency of "why!?" when developers are reading the code.

updated and pushed.

Reply:
P.S. i realized that what we need is but the first matched image. so added a cleanup change to reflect my findings.